File: src/sdks/dotnet/Application/MapFileParser.cs
namespace Envilder.Application;

using global::Envilder.Domain;
using System.Collections.Generic;
using System.Text.Json;
using System.Text.Json.Serialization;

/// <summary>
/// Parses a JSON map file into a <see cref="ParsedMapFile"/> containing
/// provider configuration (<c>$config</c>) and environment variable mappings.
/// </summary>
public class MapFileParser
{
	private const string ConfigKey = "$config";

	private static readonly JsonSerializerOptions SerializerOptions = new()
	{
		PropertyNameCaseInsensitive = true,
		Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) },
	};

	/// <summary>
	/// Parses raw JSON into a <see cref="ParsedMapFile"/>.
	/// The optional <c>$config</c> object is extracted as <see cref="MapFileConfig"/>;
	/// all other top-level string properties become secret mappings.
	/// </summary>
	/// <param name="json">Raw JSON content of the map file.</param>
	/// <returns>A <see cref="ParsedMapFile"/> ready for secret resolution.</returns>
	public ParsedMapFile Parse(string json)
	{
		using var document = JsonDocument.Parse(json);
		var mappings = new Dictionary<string, string>();
		var config = new MapFileConfig();

		foreach (var property in document.RootElement.EnumerateObject())
		{
			if (property.Name == ConfigKey)
			{
				if (property.Value.ValueKind == JsonValueKind.Object)
				{
					config = JsonSerializer.Deserialize<MapFileConfig>(property.Value.GetRawText(), SerializerOptions)
						?? new();
				}

				continue;
			}

			if (property.Value.ValueKind != JsonValueKind.String)
			{
				continue;
			}

			mappings[property.Name] = property.Value.GetString()!;
		}

		return new(config, mappings);
	}
}

File: src/sdks/typescript/src/application/map-file-parser.ts
import type { MapFileConfig } from '../domain/map-file-config.js';
import type { ParsedMapFile } from '../domain/parsed-map-file.js';
import { SecretProviderType } from '../domain/secret-provider-type.js';

const CONFIG_KEY = '$config';

const PROVIDER_MAP: Record<string, SecretProviderType> = {
  aws: SecretProviderType.Aws,
  azure: SecretProviderType.Azure,
};

/**
 * Parses a JSON map-file string into a {@link ParsedMapFile}.
 */
export class MapFileParser {
  /**
   * Parse a JSON map-file string, extracting `$config` and variable mappings.
   *
   * @param json - Raw JSON string of the map file.
   * @returns Parsed config and variable mappings.
   */
  parse(json: string): ParsedMapFile {
    const raw = JSON.parse(json) as Record<string, unknown>;
    const mappings = new Map<string, string>();
    let config: MapFileConfig = {};

    for (const [key, value] of Object.entries(raw)) {
      if (key === CONFIG_KEY && typeof value === 'object' && value !== null) {
        const obj = value as Record<string, unknown>;
        const providerStr =
          typeof obj.provider === 'string'
            ? obj.provider.toLowerCase()
            : undefined;
        config = {
          provider: providerStr ? PROVIDER_MAP[providerStr] : undefined,
          vaultUrl: typeof obj.vaultUrl === 'string' ? obj.vaultUrl : undefined,
          profile: typeof obj.profile === 'string' ? obj.profile : undefined,
        };
      } else if (typeof value === 'string') {
        mappings.set(key, value);
      }
    }

    return { config, mappings };
  }
}
